### PR TITLE
[WIP] - [SDN-2482] API enhancement to implement CR for External Gateway

### PR DESCRIPTION
**Pull request by jordigilh**
This PR aims to implement the external gateway part of the enhancement defined [here](https://github.com/openshift/enhancements/pull/1338).

It's a Work In Progress.

@trozet @tssurya PTAL.

### CODE REVIEW

File: go-controller/pkg/ovn/controller/apbroute/external_controller_pod.go [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+	if oldPod.ResourceVersion == newPod.ResourceVersion ||
+		!newPod.GetDeletionTimestamp().IsZero() {
+		return
+	}
+
+	c.podQueue.Add(
+		handlerObj{
+			op:     updateOp,
+			oldObj: oldObj,
```
Comment (trozet) [inline go-controller/pkg/ovn/controller/apbroute/external_controller_pod.go:35]: i see with the handlerObj you are trying to store the event that caused this key to be added for update by including the old and new obj. We are trying to move to a more level driven approach where we are just notified a key changed and then we resync everything for that key. I guess you want the old one to solve these cases right?

```
// processUpdatePod has to tackle a different set of scenarios:
// case 1: Old and new pods are gateways: Need to update the new Pod IP to all impacted namespaces. First delete the old one and then add the new one
// case 2: Old is a gateway but the new is not: Delete the old pod's IP from all impacted namespaces
// case 3: Old is not a gateway but the new one is: Add the new pod's IP to all impacted namespaces defined by the matching policies
// case 4: Old and new belong to a namespace impacted by a policy: Update the logical routes to the pod as defined by the external routes.
// case 5: Old and new are not gateways and are not impacted by a policy: nothing to do
```

will try to think about how we can get rid of this dependency

Comment (jordigilh) [reply go-controller/pkg/ovn/controller/apbroute/external_controller_pod.go:35]: Correct. At the moment this approach makes it easy to identify the differences. I guess what you mean is to recalculate the expected gateway info from the changes brought to the pod and apply the difference with what's already there, and disregard the operation performed.

I will try to change it to this model in this PR if that follows the expected approach.

Comment (trozet) [reply go-controller/pkg/ovn/controller/apbroute/external_controller_pod.go:35]: // processUpdatePod has to tackle a different set of scenarios:
// case 1: Old and new pods are gateways: Need to update the new Pod IP to all impacted namespaces. First delete the old one and then add the new one

I don't think the pod IP can change after a pod has been added. In that case we dont need to worry about this case right?

// case 2: Old is a gateway but the new is not: Delete the old pod's IP from all impacted namespaces

In this case you could keep an internal cache to know if this pod was being tracked previously as a gateway. You don't need the old object here.

// case 3: Old is not a gateway but the new one is: Add the new pod's IP to all impacted namespaces defined by the matching policies

This is the same as adding a new pod.

// case 4: Old and new belong to a namespace impacted by a policy: Update the logical routes to the pod as defined by the external routes.

The ips shouldn't change so what else do we need for this case?

// case 5: Old and new are not gateways and are not impacted by a policy: nothing to do

Comment (jordigilh) [reply go-controller/pkg/ovn/controller/apbroute/external_controller_pod.go:35]: >case 1: Old and new pods are gateways: Need to update the new Pod IP to all impacted namespaces. First delete the old one and then add the new one
> 
> I don't think the pod IP can change after a pod has been added. In that case we dont need to worry about this case right?

Ah I see your point. Yes you are right: the update process will not include killing the pod and restarting a new one, like for instance when the whole pod dies and a new one is respawned. In that case we would be covered by the delete and subsequent add operations.

I see the update only when there are changes to the pod without killing it, such as interacting with labels, annotations or even the status spec.

I will redo this part. Thanks!

> 
> // case 2: Old is a gateway but the new is not: Delete the old pod's IP from all impacted namespaces
> 
> In this case you could keep an internal cache to know if this pod was being tracked previously as a gateway. You don't need the old object here.

Agreed, I just started tracking all this in an internal cache while revisiting the handlers for namespace.

> // case 3: Old is not a gateway but the new one is: Add the new pod's IP to all impacted namespaces defined by the matching policies
> 
> This is the same as adding a new pod.

Agreed. 

> 
> // case 4: Old and new belong to a namespace impacted by a policy: Update the logical routes to the pod as defined by the external routes.
> 
> The ips shouldn't change so what else do we need for this case?
> 

Yes, you're right: the IPs should not change in this case, so this is not going to happen.

**Comment by jordigilh** [comment]
Will create a new PR since I closed this one to allow me to continue working without triggering notifications for each commit I made during development.

**Pull request closed without merging**